client/src/main/java/io/confluent/kafka/schemaregistry/client/rest/RestService.java

```diff
@@ -526,9 +528,14 @@ private void setBasicAuthRequestHeader(HttpURLConnection connection) {
     String userInfo;
     if (basicAuthCredentialProvider != null
         && (userInfo = basicAuthCredentialProvider.getUserInfo(connection.getURL())) != null) {
-      byte[] userInfoBytes = userInfo.getBytes(StandardCharsets.UTF_8);
-      String authHeader = DatatypeConverter.printBase64Binary(userInfoBytes);
-      connection.setRequestProperty("Authorization", "Basic " + authHeader);
+      try {
+        byte[] userInfoBytes =
+          URLDecoder.decode(userInfo, "UTF-8").getBytes(StandardCharsets.UTF_8);
+        String authHeader = DatatypeConverter.printBase64Binary(userInfoBytes);
+        connection.setRequestProperty("Authorization", "Basic " + authHeader);
+      } catch (UnsupportedEncodingException e) {
+        return;
```

Comment:
We shouldn't swallow this exception. Looks like we could wrap it in an IllegalArgumentException so it gets passed back to the user (like in parseBaseUrl).

Reply:
Yeah, I wouldn't wrap the lines here in a try-catch block. We want to indicate to the calling methods that an error happened.

Reply:
Uhm, I’l gladly change it, but the UnsupportedEncodingException is thrown, if "UTF-8" is not supported by the stdlib/jvm as possible encoding, so IllegalArgumentExceptions seems like an odd choice to me. I think usually one would throw an AssertionError in this case. If the encoding is supported or not depends on the environment rather than what the user provides as data.

Reply:
Good point, I'm less opinionated on the exception thrown or whether we log something. Just know it'll be a pain for people to troubleshoot if we silently swallow it. @wicknicks ?

Reply:
Yes @codyaray, definitely agree that we shouldn't simply swallow the exception here. Good point, @i7c! Also, since this method is called `setBasicAuthRequestHeader`, I am less inclined to do the decoding here. The more appropriate place would be just before calling this method on line 157 in this class, and have the setter only set stuff, something like this: 

```
if (basicAuthCredentialProvider != null) {
  userInfo = basicAuthCredentialProvider.getUserInfo(connection.getURL()))
  userInfoBytes = URLDecoder.decode(userInfo, "UTF-8").getBytes(StandardCharsets.UTF_8);
  setBasicAuthRequestHeader(userInfoBytes, connection);
}
```

This would also allow easier handling of the IOException. The other alternative is to have `setBasicAuthRequestHeader` throw an IOException, which is a bit odd as well.

Reply:
First of all thanks for the helpful feedback. I get the points and I agree with what you said.

However, trying out your proposal bumped up cyclomatic and NPath complexity of sendHttpRequest() above the allowed limits, which made me think: Why should this method (or the setter) have to deal with decoding user input anyways?

Wouldn't it make more sense to have the CredentialProviders deal with decoding? This would have some nice properties:
* we can decode only once (when BasicAuthCredentialProvider.config() is called or in a "caching manner" in case of the UrlBasicAuthCredentialProvider), so we reduce execution time of sendHttpRequest()
* we can deal with errors where they occur conceptually, in "providing credentials"
* we can offer a nicer way for the user to specify credentials (with a new Provider implementation), namely with two properties "username" and "password" where no URL encoding is necessary. I actually don't like the idea that the user has to put something like johndoe:my%20secret in there anyways. It also might cause some hassle when using configuration management. Think of a centralised store for secrets, so configuration management would need to URL-encode the provided secret first ...)

Please let me know what you think about it and if this is getting too big for a simple feature. ;)
I’ll gladly provide a patch, should you like the idea. :smile:

Reply:
I like the direction you're thinking in. We actually already do support passing user and password separately. Seems like it may just be undocumented/poorly documented. So I'm not sure how far down this road with code refactoring versus docs improvements.

Here's a gist that does SR auth using system properties.
https://gist.github.com/mageshn/c623f2d43b66b1153f5da32662ca5e94

(Sorry if this possibility was already obvious to you. This isn't a code base I normally touch.)

Reply:
@i7c: thanks a lot for your thoughts on this. Would love that patch. Thank you! 

Some thoughts: I like the new contract being established in your second suggestion. 

Also, maybe we can create an abstract class which extends `BasicAuthCredentialProvider` to house your suggestions. The other Provider implementations would just extend this new class. Once we officially move to Java 8, this could be replaced with a default method in the interface.